### PR TITLE
feat: add get prover data api function

### PR DESCRIPTION
**Pull request by atanmarko**
# Description

Implement jsonrpc `eth_getProverData` api function that returns account data used in the block.

Fixes TP-612

## Additions and Changes
* JSONRPC interface expanded with `eth_getProverData` function
* Added prover module for implementation of data retrieval and filtering functions
* Internal APIs are updated to run the tracer, retrieve merkle proof for the contract storage and state

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: state/immutable-trie/snapshot.go [hunk 1]
```diff
@@ -53,6 +53,36 @@ func (s *Snapshot) GetStorage(addr types.Address, root types.Hash, rawkey types.
...
+
+	proof, ok := trie.GetProof(key, s.state.storage)
+	if !ok {
+		return nil, nil
+	}
+
+	// Reverse merkle proof array, so that root is at the beginning
+	for i, j := 0, len(proof)-1; i < j; i, j = i+1, j-1 {
+		proof[i], proof[j] = proof[j], proof[i]
+	}
```
Comment (hadjiszs) [inline state/immutable-trie/snapshot.go:81]: no function in the lib to reverse an array?

File: state/immutable-trie/snapshot.go [hunk 2]
```diff
@@ -69,6 +99,22 @@ func (s *Snapshot) GetAccount(addr types.Address) (*state.Account, error) {
...
+
+	proof, ok := s.trie.GetProof(key, s.state.storage)
+	if !ok {
+		return nil, nil
+	}
+
+	// Reverse merkle proof array, so that root is at the beginning
+	for i, j := 0, len(proof)-1; i < j; i, j = i+1, j-1 {
+		proof[i], proof[j] = proof[j], proof[i]
+	}
```
Comment (hadjiszs) [inline state/immutable-trie/snapshot.go:113]: same here

File: server/server.go [hunk 3]
```diff
@@ -612,6 +638,25 @@ func (j *jsonRPCHub) GetStorage(stateRoot types.Hash, addr types.Address, slot t
 	return res.Bytes(), nil
 }
 
+func (j *jsonRPCHub) GetStorageProof(stateRoot types.Hash, addr types.Address, slot types.Hash) ([][]byte, error) {
```
Comment (hadjiszs) [inline server/server.go:643]: maybe a TODO or a simple NOTE to inform switching to the standard API function once it is available?

**Review by hadjiszs — APPROVED** [review]

File: state/immutable-trie/snapshot.go (hunk 1 above)
Comment (atanmarko) [reply state/immutable-trie/snapshot.go:81]: No, was googling for it

**Pull request merged**